### DISCUSSION TITLE
Allow calling base class constructor in the middle of derived class constructor

### DISCUSSION THREAD

**Discussion by munael** [dotnet/csharplang#4827] [Language Ideas]
To allow

```cs
class Derived : Base {
    Derived(Foo foo) {
        // do bar and baz with foo
        base(bar, baz);
        //  blah
    }
}
```

Instead of having to extract the `do bar and baz` section into static functions (one for each argument in the base constructor that needs modified values). If we can already functionally do this, there should be no need for the extra complication.

Any standing _theoretical_ reasons for the restriction?

**Comment by HaloFour** [comment]
@narfanar 

> Any standing reasons for the restriction?

It's easier to comprehend that you're very limited as to what you can do before the base constructor call when the syntax forces it on you.  Yes, you can make your code slightly less verbose by eliminating the need for a separate function, but it's a suitable workaround for what is not a terribly common situation.

**Comment by CyrusNajmabadi** [comment]
> Any standing reasons for the restriction?

Yes, the language doesn't allow it.

As for why it doesn't allow it, the general reason is around the complexity of the programming model, especially around initialization, virtual-calls, and whatnot.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> As for why it doesn't allow it, the general reason is around the complexity of the programming model, especially around initialization, virtual-calls, and whatnot.

I get the sense from the question that they're not looking to ease the restrictions, only allow the restricted code to exist before the base constructor call.

**Comment by munael** [comment]
> @narfanar
> 
> > Any standing reasons for the restriction?
> 
> It's easier to comprehend that you're very limited as to what you can do before the base constructor call when the syntax forces it on you. Yes, you can make your code slightly less verbose by eliminating the need for a separate function, but it's a suitable workaround for what is not a terribly common situation.

The verbosity (and even wasted work) rapidly spiral out of control once several arguments of the base class constructor need to be themselves prepared from the arguments of the derived class constructor.

Verbosity because then we need separate static functions (and calls) to get the value of each argument passed to the base constructor, and wasted work if it happens that several such arguments can share parts of their preparation functions.

D allows this. I remember there were issues with virtual functions and some order of construction stuff. Will see if there're salient lessons there that could help the case of this issue.

**Reply by voroninp** [reply]
IL also allows it, btw.

```
public class Foo
{
	public Foo(int i)
	{		
	}
}

public class Bar: Foo
{
	public Bar(string s):base(int.Parse(s))
	{		
	}
}
```

And here's how `Bar` constructor looks like:
```
.method public hidebysig specialname rtspecialname 
	instance void .ctor (
		string s
	) cil managed 
{
	.custom instance void System.Runtime.CompilerServices.NullableContextAttribute::.ctor(uint8) = (
		01 00 01 00 00
	)
	// Method begins at RVA 0x2153
	// Header size: 1
	// Code size: 15 (0xf)
	.maxstack 8

	// {
	IL_0000: ldarg.0
	IL_0001: ldarg.1
	IL_0002: call int32 [System.Runtime]System.Int32::Parse(string)
	// (no C# code)
	IL_0007: call instance void UserQuery/Foo::.ctor(int32)
	IL_000c: nop
	// }
	IL_000d: nop
	IL_000e: ret
} // end of method Bar::.ctor
```

So yes, C# should also allow this:
```
public Bar(string x, string y)
{
    var xi = int.Parse(x);
    var yi = int.Parse(y);
    base(xi+yi);
}
}
```

**Comment by YairHalberstadt** [comment]
> and wasted work if it happens that several such arguments can share parts of their preparation functions.

Whilst a hacky workaround, you could use out parameters to avoid this wasted work.

**Comment by dsaf** [comment]
@narfanar 

> Any standing *theoretical* reasons for the restriction?

https://en.wikipedia.org/wiki/Halting_problem

**Comment by munael** [comment]
@YairHalberstadt 

How would you out into the arguments of the base constructor? Can you provide an example?

@dsaf 
> @narfanar
> 
> > Any standing _theoretical_ reasons for the restriction?
> 
> https://en.wikipedia.org/wiki/Halting_problem

Sorry? I'm not following.

**Comment by YairHalberstadt** [comment]
```csharp
Public class A
{
    public A(int a, int b) {}
}
Public class B : A
{
    public B(string str) : base(GetBaseParams(str out int b), b){}
    public int GetBaseParams (string str, out int b)
    {
         strs = string.Split(str);
         b = int.Parse(strs[1]);
         return int.Parse(strs[0]);
    }
}

**Reply by yahoo0742** [reply]
@YairHalberstadt If the `GetBaseParams` is static from your example, I'd support it. Otherwise I prefer the to receive a warning.

**Comment by The-Futurist** [comment]
@CyrusNajmabadi 

> > Any standing reasons for the restriction?
> 
> Yes, the language doesn't allow it.

Another vacuous reply from Cyrus! Given that his question literally _means_ "Any standing reasons for the language disallowing it?" I can only conclude that exaggerated self-importance often motivates much of what your say here.

**Comment by yaakov-h** [comment]
@Korporal and then Cyrus addresses the reasons in literally the next line. Personal attacks are not appreciated here.

**Comment by HaloFour** [comment]
I'm kind of on the fence about this.  I have written classes that had to manipulate/validate parameters before passing them along to a base constructor before, pre-C# 7.0, and it was painful as all Heck.  With `out` parameters it's considerably less painful since you can pass the arguments along to a `static` method that does the work and passes them back, but that's still not exactly pleasant and it moves the initialization code far away from where it's relevant.  It seems that it should be safe to allow arbitrary statements before `base(...)` as long as `this` is inaccessible (including all instance members).  IIRC that's what the CLR allows, it's just that the current C# syntax doesn't allow anywhere to put those statements.

There could be two ways of accomplishing this syntax-wise.  You could have a block before the base constructor call:

```cs
private readonly int y;

public Derived(string x, int y) {
    // static stuff here only
    if (string.IsNullOrEmpty(x)) throw ArgumentException();
} : base(x) {
    // can access instance members here
    this.y = y;
}
```

Or allow `base` to appear like a method call within the constructor body:

```cs
private readonly int y;

public Derived(string x, int y) {
    // static stuff here only
    if (string.IsNullOrEmpty(x)) throw ArgumentException();
    base(x);
    // can access instance members here
    this.y = y;
}
```

The first trades some verbosity for very clear delineation of the sections of code.

**Comment by YairHalberstadt** [comment]
I'm leaning towards the first option, but http://gafter.blogspot.com/2017/06/making-new-language-features-stand-out.html?m=1

**Comment by theunrepentantgeek** [comment]
I cut my teeth as a professional developer using Delphi - and its Object Pascal explicitly allowed/required a call to the base constructor as a regular statement.

It was _common_ to find subtle object initialization bugs where certain instance methods would only work _after_ the base constructor call. Sometimes, when we were lucky, they'd throw. Other times, they'd return incorrect or inconsistent results.

I vividly remember having to untangle some chicken-and-egg scenarios where function `Foo()` would only give correct results after the base constructor was called, but the result of `Foo()` was needed to make that call.

**Comment by CyrusNajmabadi** [comment]
>Another vacuous reply from Cyrus! 

I literally explained why in the post you snipped from and didn't quote fully.  The language doesn't allow this because of "the general reason [...] around the complexity of the programming model, especially around initialization, virtual-calls, and whatnot."

That is truthful and accurate and appropriate given the question.

**Comment by The-Futurist** [comment]
@CyrusNajmabadi 

> > Another vacuous reply from Cyrus!
> 
> I literally explained why in the post you snipped from and didn't quote fully. The language doesn't allow this because of "the general reason [...] around the complexity of the programming model, especially around initialization, virtual-calls, and whatnot."
> 
> That is truthful and accurate and appropriate given the question.

Well Cyrus I'm afraid I've found you to be somewhat supercilious in many of your exchanges, a person asks "Any standing reasons for the restriction?" and you simply cannot resist trying to appear erudite by responding "Yes, the language doesn't allow it." which is a tautology.

Most of us would have omitted that and simply said the second part of your response. You must understand that the tone and character one uses in these posts is an important aspect of communication be it technical or otherwise, I simply think you need some work in this area.

**Comment by CyrusNajmabadi** [comment]
> and you simply cannot resist trying to appear erudite by responding "Yes, the language doesn't allow it." which is a tautology.

I literally gave the reason in the *very next sentence*.  It was one full thought here.  I included thefirst part because sometimes people view these issues as potentially bugs in the compiler and not as per the exact and specific design of the language.  I was stating emphatically that this was an intentional decision in the language, and i then followed *immediately* with teh reasoning for why it was an intentional decision.

It is important as part of the discussion because these can be separate issues with differing rules about what is done with them.  For example, if this was *not* per the language rules (which does happen), we could talk about what we should then do about that.  But if this is per the language rule, then this becomes a discussion about *why* it is that way.  I intentionally specified that this is by design, and the gave information on why the design is the way it is.

> Most of us would have omitted that

And i would not.  Because it's important to recognize the intentional decisions from the unintentional ones *especially* in the domain of language design.

**Comment by HaloFour** [comment]
To argue against myself it's also just as easy to use a `private` constructor with a `static` factory method:

```cs
public class A
{
    public A(int a, int b) {}
}
public class B : A
{
    private B(string a, int b) : base(a, b)
    {
        // more init here
    }

    public static B Create(string str)
    {
        // validation and pre-init here 
         var strs = string.Split(str);
         int b = int.Parse(strs[1]);
         var instance = new B(strs[0], b);
        // post-init here, if necessary
        return instance;
    }
}
```

**Comment by juliusfriedman** [comment]
Please show a real life example of how this is useful as the underlying base constructor must be called anyhow.. 

For the layman technica you can't pass it around as the base instance until you do so and that base may still throw an exception; furthermore where are you going to get allocated when you define the class? so I am baffled by this proposal.. unless it's for custom allocator semantics which allow control over such...

You would / could use GetUnitializedObject and composition to achieve this now as well as unsafe code but I still don't understand why you would want to do this as it's asking for problems.

**Comment by HaloFour** [comment]
@juliusfriedman 

The C# compiler does this all the time, in that you're allowed to have expressions in the `base()` call.  That's compiled into IL that executes before the base constructor call, and as long as that IL doesn't attempt to access `this` it's perfectly fine.  Where it becomes a problem is that you can only have expressions, and those expressions can only have very limited interaction with one another.  So if you want to do something like validate that the constructor arguments are valid before calling the base constructor you have to extract that out into a separate static method which returns a value that you can pass to the base constructor.

**Comment by juliusfriedman** [comment]
Yes but shouldn't the base class have that responsibility i.e. if it changes in the future...

Also since the interaction is limited to self reference but explicitly bad when accessing the base Pointer itself then what's the use?

So long as this cannot be used I think you mean from dervied as they check with call virt if call virt is not used this is still fine but only on the members defined on this in that type...

Tldr, how can this be useful and not bad?

Now with what I was saying about constraints being utilized in the constructor via modreqs.. I see that enhancing generics such as to support ctor arguments would be needed for this as well .. https://github.com/dotnet/csharplang/issues/769

**Comment by HaloFour** [comment]
@juliusfriedman 

> Yes but shouldn't the base class have that responsibility i.e. if it changes in the future...

Why should the base class be at all aware of what a derived constructor might want to do?

> Also since the interaction is limited to self reference but explicitly bad when accessing the base Pointer itself then what's the use?

The interaction would be limited to static members of the type as well as the constructor arguments themselves.  The use would be to do something with those constructor arguments before calling the base constructor, such as validating them.

> Tldr, how can this be useful and not bad?

An example would be wanting to validate that the constructor arguments are valid before calling a base constructor that might do something expensive.  Another example would be deriving the base constructor arguments from the derived constructor arguments through some calculation.  Anything that would be nicer to express as statements rather than trying to pack into one or more isolated expressions, e.g.:

```cs
public class Derived : Base {
    private string value;
    public Derived(string value) {
        if (value == null) throw new ArgumentNullException(nameof(value));

        // base constructor does something very expensive
        // that you want to avoid if the arguments aren't valid
        base();
        this.value = value;
    }
}
```

And I'll note again that all of this is perfectly legal as far as the CLR is concerned, and C# takes advantage of this fact by virtue of allowing you to execute expressions to derive the base constructor arguments.  The limitation against invoking other arbitrary code (that doesn't attempt to access `this`) is imposed by the C# compiler.

**Comment by juliusfriedman** [comment]
@HaloFour

It sounds like a violation of the chain of responsibility, the base class should validate its own parameters, ifcase then they change the validate logic is then wrong.

I have wanted to do that for reasons you cite such as validation of parameters but it usually ends up in duplicate code with no benefit to the user or maintainer, what is more usual is adjusting the parameters passed to base in such conditional statenents such as null being pass as empty or -1 implying to check another parameters like length and then defaulting to empty if null etc which again should probably be handled in the base class.

Tldr...

Finally, how will you prevent double initialization here? E.g. if I call base twice or with different arguments each time etc .. what if I call it in a different thread? What about guarantees about the construction in a static context..

So much to consider and adjust for with little demonstrated benefit ..

Qi Bono?

Tody you can just pass an action\task to the Ctor which can do this without the side effects and maintains safety until you shoot yourself in the foot.

This feature will give everyone a loaded gun...

**Comment by HaloFour** [comment]
@juliusfriedman 

> the base class should validate its own parameters

Yes, the base class should validate its own parameters, just like the derived class should validate its own parameters.  The problem is that, currently, the derived class can't validate its own parameters until after the base class constructor is invoked, unless you contort your logic into expressions that can be passed to the base class.  If the base class accepts no parameters, you're completely out of luck.

**Comment by juliusfriedman** [comment]
@HaloFour

In what situation would that be useful, if the property isnt virtual you cannot access it from the derived until you call base ctor.

If your base ctor accepts no arguments and you must validate something on it before you construct it then favor composition over inheritance for the poorly design base class. 

That is a classic case of a paradox, you can't validate something which doesn't exist until after you create it.

In the extent that the object instance requires the state of some static type or a Singleton which is aggravated by access then reflection or Emit becomes the conical way of dealing with such AFAIK.

**Comment by HaloFour** [comment]
@juliusfriedman 

> This feature will give everyone a loaded gun...

I don't understand why you think this.  Maybe we're not on the same page?  As mentioned, you're free to do this already in C# in that you can execute arbitrary expressions and pass their results to the base constructor arguments.

> In what situation would that be useful, if the property isnt virtual you cannot access it from the derived until you call base ctor.

Nobody is talking about properties.  I'm talking explicitly and exclusively about the parameters to the derived constructor.

> That is a classic case of a paradox, you can't validate something which doesn't exist until after you create it.

Again, I'm not talking about validating the object which you are trying to construct before you've constructed it.

**Comment by HaloFour** [comment]
@juliusfriedman 

> Finally, how will you prevent double initialization here? E.g. if I call base twice or with different arguments each time etc .. what if I call it in a different thread?

This is indeed the only tricky aspect about a feature like this.  The compiler would have to enforce that you invoke the `base` constructor *exactly once* within that method and that the only thing that you can do with the `base` constructor is to invoke it.

**Comment by juliusfriedman** [comment]
@HaloFour, see above...

Double initialization; different initialization; both combined with some type of threading and then used as a static instance.

It's one thing to have the construct in IL which is not the topic of discussion here.

My problem would be letting the compiler do this as again what benefit is there besides chaos which can and will likely ensure this becomes more aggravated than it is already.

Furthermore the only use would be a niche object which had a contract that it would perform different logic in different constructors and would need both to be called, the solution to that is as you have provided above in the example, write an init method or two and call them as needed.

Finally structs also have a consideration as I have to assign all members before return or call base first... if struct inheritance is ever championed that could change but still a consideration.

**Comment by HaloFour** [comment]
@juliusfriedman 

> My problem would be letting the compiler do this as again what benefit is there besides chaos which can and will likely ensure this becomes more aggravated than it is already.

There is no chaos here.  You're reading something into this proposal that isn't there.

> Furthermore the only use would be a niche object which had a contract that it would perform different logic in different constructors and would need both to be called, the solution to that is as you have provided above in the example, write an init method or two and call them as needed.

Derived classes may always impose a different contract on construction from the base class.  That's why they can have their own constructors.  The derived constructor is purely additive, though, and this proposal does nothing to change that.  You can't change how the base constructor works, nor can you avoid invoking it (short of throwing an exception).

Sure, you can achieve the same thing with an init method, and I mentioned that above.  But the one issue that an init method brings is that it makes it more awkward to extend from the derived class as you can no longer chain constructors.

As for whether or not this is niche, I'd suggest that it's not niche to want to validate arguments early, nor is it niche for a derived class to expose different constructor parameters than the base class.

**Comment by HaloFour** [comment]
@juliusfriedman 

Here's an example of how you can workaround the language limitation today, although it only works if there is an overload of the base constructor that accepts arguments:

```cs
public class Base {
    protected Base(int a, string b, Guid c) => Thread.Sleep(TimeSpan.Infinity);
}

public class Derived : Base {
    public Derived(SomeType t) : base(PreInitialize(t, out var b, out var c), b, c) {
        // other init here
    }

    private static int PreInitialize(SomeType t, out string b, out Guid c) {
        if (t == null) throw new ArgumentNullException(nameof(t));
        int a = CalculateA(t);
        if (a < 0) throw new InvalidOperationException();
        b = CalculateB(t, a);
        c = CalculateC(t, a, b);
        return a;
    }
}
```

Before `out`-declarations this was a *lot* more awkward to write since you couldn't otherwise declare those variables, so you needed three separate static methods which would likely involve code duplication.

**Comment by juliusfriedman** [comment]
@HaloFour 

Could you also not have a ParameterValidator T and do this in an OOP way...

There's no reason to divorce the construction process to support parameter validation iMHO.

What other languages support this? JavaScript?

What usefulness will this lend to existing paradims? Besides the contrived..

Tldr;

I appreciate your willingness to think out of the box!

I will take back chaos by saying it's good you will be able to now express something you could in IL from C# however care should be taken to ensure that we don't open a can of worms and that the effort expended to champion the proposal will be outweighed by the benefit it provides which iMHO has not been shown.

Tldr; I don't want this to result in reinit. Especially with structs because assignments made before that point to the instance will be erased by the init call to this unless you expense more energy to ensure that the enhancement doesn't work for structs unless before assigned.

**Comment by HaloFour** [comment]
@juliusfriedman 

> Why not have a ParameterValidator T and do this in an OOP way...

That adds more complexity and moves the logic away from where it is relevant.

> There's no reason to divorce the construction process to support parameter validation iMHO.

It doesn't divorce the construction process, it only enables more code constructs to exist before the base constructor call.

> What other languages support this? JavaScript?

D does.  Check out [15.8.9](https://dlang.org/spec/class.html#constructors) of the D spec.  In particular:

> 2. If a constructor's code contains a delegate constructor call, all possible execution paths through the constructor must make exactly one delegate constructor call:
> 3. It is illegal to refer to `this` implicitly or explicitly prior to making a delegate constructor call.

In D the "delegate constructor" is the base constructor.

I'm not aware of other OOP languages that the developer to explicitly do this, although I think all OOP languages allow some code to execute before the base constructor call for field initializers as well as the previously mentioned expressions evaluated for the base constructor arguments.

> I will take back chaos by saying it's good you will be able to now express something you could in IL from C# however care should be taken to ensure that we don't open a can of worms and that the effort expended to champion the proposal will be outweighed by the benefit it provides which iMHO has not been shown.

Sure, I'm not going to argue against that point.  Of course it also applies to literally every other proposal here.  😁

**Comment by theunrepentantgeek** [comment]
> What other languages support this?

I don't know about the modern versions, but Delphi supported this at least in versions 1-7 of its Object Pascal language. IIRC, the compiler enforced that exactly one inherited constructor call was made.

As I recall, being able to run code prior to invoking the constructor of the base class was immensely useful in some situations - though it's been so long that I'm fuzzy on the details. I *do* remember the C# style being very constraining when I was first picking up the language though ...

**Comment by HaloFour** [comment]
@theunrepentantgeek 

> I don't know about the modern versions, but Delphi supported this at least in versions 1-7 of its Object Pascal language. IIRC, the compiler enforced that exactly one inherited constructor call was made.

Looks like that's still the case with Oxygene:

https://docs.elementscompiler.com/Oxygene/Members/Constructors/

**Comment by yaakov-h** [comment]
Here's a real-world use-case I came across just last week:

```csharp
using System;

namespace MyLibrary
{
    public partial interface IRuleInfo
    {
        public Guid Id { get; }
    }

    [Serializable]
    public class RuleValidationException : Exception
    {
        public RuleValidationException(IRuleInfo ruleInfo, Exception inner)
            : base(GenerateMessage(ruleInfo), inner)
        {
        }

        /* Additional constructors omitted */

        static string GenerateMessage(IRuleInfo ruleInfo = null)
        {
            return ruleInfo == null ? "Rule validation failed" : "Rule validation failed for Rule id: " + ruleInfo.Id;
        }
    }
}

```

It would be a lot nicer and easier to read if I could use the following constructor instead:

```csharp
public RuleValidationException(IRuleInfo ruleInfo, Exception inner)
{
    if (ruleInfo == null)
    {
        base("Rule validation failed.", inner);
    }
    else
    {
        base("Rule validation failed for Rule id: " + ruleInfo.Id, inner);
    }
}
```

**Comment by juliusfriedman** [comment]
@yaakov-h, I👏your efforts but find your first example better as it encapsulates OOP.

You can return an exception from a Validate method and not throw it too but that's rather heavy... why not just have a Validate which returns a State perhaps with a dictionary or otherwise...

The example you provided cannot be overcome by the base class changing the constructor which it calls and or another exception being thrown from that constructor anyway.

Also you can't really even guarantee that your constructor won't encounter the oom due to base classes logic so a Validate or Initialize method will beat this hands down evrtytime and doesnt require the effort to implement nor provide the issues which this may especially for structs.

**Comment by HaloFour** [comment]
@juliusfriedman 

> find your first example better as it encapsulates OOP.

What does that have to do with OOP?  That's like arguing that you can never override a member unless the very first thing you do in the overridden method is to call the base method.  The fact that you can do other things before invoking the base method doesn't break anything about OOP.  Heck, the fact that you're allowed to invoke the base method zero times, or a thousand times, doesn't break OOP.  There's no tenet of OOP that requires that a base constructor call be the first "statement".

> The example you provided cannot be overcome by the base class changing the constructor which it calls and or another exception being thrown from that constructor anyway.

That doesn't make sense.  The derived class calls the base constructor in this example.  If the base class was changed in that the constructor was no longer compatible that would break all code consuming the base class, irrespective of this proposal.

> Also you can't really even guarantee that your constructor won't encounter the oom due to base classes logic so a Validate or Initialize method will beat this hands down evrtytime and doesnt require the effort to implement nor provide the issues which this may especially for structs.

Nobody is asking for such a guarantee.  Nothing can guarantee that a field initializer won't throw, or that a static method called by the derived constructor within the base constructor call won't throw.  Forcing the developer to move otherwise local validation and calculation code doesn't change any of that equation at all, except to break code locality and force additional boilerplate.

**Comment by juliusfriedman** [comment]
@HaloFour

constructor which it calls, is what I said, all objects call a base constructor either ValueType or object or otherwise as required by the CLR.

The difference with encapsulation is overhead, I can choose and detect if I am validated either with Memoization or otherwise (caching).

It allows that portion of code to give it's own implementation if desired i.e. an IValidator which is a better encapsulation of multiple principles of OOP. 

It also changes the state of a struct although I suppose you can already assign to this with structs in most cases and with classes through initialization and now that you mention static as well in the CLR due to state and guaranteed made about such which will NOT only be difficult to change now for everyone but can for that matter already be handled by the most determined either with a Host or Hijacking for better yet proper programming techniques which dont require us to do this in the first place.

Now for that matter I vote for this! 

I'd in alot ways rather see this championed than default interface implementations...

Finally for that matter I would absokutely love to this feature on partial classes!

**Comment by yaakov-h** [comment]
*[scratches head]*

I don't get how any of that is relevant at all...

**Comment by munael** [comment]
Some restrictions that would allow some new constructs but might be easier to defend:

Example of allowed uses:
```cs
class Derived : Base {
    Derived(Foo foo) {
        // do bar and baz with foo
        base(bar, baz);
        //  blah
    }
}
```
1. Exactly one `base` constructor call in the `Derived` constructor call is allowed.
2. The `base` constructor call must have at least one argument (`base(T)`) (cannot defer a `base()` call).
3. The `base` call _must_ be in the outermost scope of the `Derived` constructor body. So, you cannot put it in branches or loops.
  a. How about `try` blocks?
  b. `if`/`switch` statements with an exhaustiveness check (all branches call the base at least once)?
      1. This's basically asking for linear typing for the `base` constructor, which is implicitly passed along to other calls. So, for example, you satisfy the requirement for the exactly-one call by calling a different derived class constructor that satisfies it.
4. The `base` call is a statement, not an expression.

**Comment by KalleOlaviNiemitalo** [comment]
Inspired by <https://github.com/dotnet/runtime/issues/46489>, it could be handy if a constructor could handle exceptions thrown by the base class constructor, a bit like a [function-try-block in C++](https://en.cppreference.com/w/cpp/language/function-try-block).

```C#
partial class VolumeStream : FileStream
{
    public VolumeStream(string path)
    {
        SafeFileHandle handle = null;
        try
        {
            handle = OpenVolume(path);
            int sectorSize = GetSectorSize(handle);
            base(handle, FileAccess.ReadWrite, sectorSize);
        }
        catch
        {
            handle?.Dispose();
            throw;
        }
    }
}
```

Sequence expressions (<https://github.com/dotnet/csharplang/issues/377>) do not support `try` statements and so would not be able to handle these exceptions.

I don't see any restriction against catching base constructor exceptions in ECMA-335. III.1.8.1.4 (Class and object initialization rules) even has a note that mentions this possibility.

After catching, the derived constructor would not be allowed to return; it would have to throw something or loop infinitely. A constructor of a value type could perhaps be allowed to assign to `this` and return, but that does not seem worth supporting because the [ValueType() constructor](https://docs.microsoft.com/dotnet/api/system.valuetype.-ctor?view=net-5.0#System_ValueType__ctor) is unlikely to throw. I don't think the derived constructor should be allowed to retry the base constructor after an exception, because the base constructor may assume that instance fields have default values, and the previous attempt may have modified them already.

The syntax could be distinguished from `base(T)` (<https://github.com/dotnet/csharplang/issues/2910>) by checking whether the closing parenthesis is followed by a dot or by a semicolon.

**Comment by KalleOlaviNiemitalo** [comment]
The example above has the problem that one would not know whether `base()` is implicitly called at the top of the constructor, without reading the rest of the constructor. It would be spooky action at a distance, like `yield return` in iterator methods. Perhaps it would be better to require something like `: explicit base` after the parameter list in the constructor declaration, to indicate that the default `: base()` does not apply and the body will explicitly call a constructor.

**Comment by KalleOlaviNiemitalo** [comment]
This would help in <https://github.com/dotnet/runtime/issues/38804>.

**Comment by shadow-cs** [comment]
A little late to the party but I offer another workaround, similar to the `out` arguments but a little more readable (to me at least) is using a constructor overload with a tuple:

```cs
  public class Base
  {
    protected Base(int a, string b, Guid c) { }
  }

  public class Derived : Base
  {
    public Derived(SomeType t) : this(PreInitialize(t))
    {
    }

    private Derived((int a, string b, Guid c) p) : base(p.a, p.b, p.c)
    {
      // other init here
    }

    private static (int a, string b, Guid c) PreInitialize(SomeType t)
    {
      if (t == null) throw new ArgumentNullException(nameof(t));
      int a = CalculateA(t);
      if (a < 0) throw new InvalidOperationException();
      string b = CalculateB(t, a);
      Guid c = CalculateC(t, a, b);
      return (a, b, c);
    }
  }
```

I'm using the same sample as @HaloFour.

Also using this technique gives you more freedom on what you can pass and what to do with values in the tuple (the overloaded ctor can also use them and store them in a field if needed which isn't possible without using the overload).

**Comment by bernd5** [comment]
@shadow-cs nice workaround

AFAIK the tuple solution would just not work for by-ref parameters and ref-like types (e.g. Span<T>) - but for by value calls it is nice.

**Reply by phizch** [reply]
Been a while, but.. The tuple could easily be replaced with a struct, or even a `ref struct`. 
It does make it way more verbose though.

**Comment by ghost** [comment]


**Reply by FaustVX** [reply]
The `age` and `name` only exist before the `base` or `this` block.
So even `base(age)` and `this(name)` isn't possible.

**Reply by ghost** [reply]
I think if the block before the ```base``` or ```this``` does not free the stack, I should be able to use ```base(age)``` or ```this(name)```.

In fact, it is equivalent to this code.
```
public class A
{
    public A(int age)
    {
        ...
    }
}

public class B : A
{
    public B(string name)
    {
        // static only before the ‘base(age)’
        ...
        int age = 10;
        ...
        base(age); // end
    }
    {
        ...
    }

    public B()
    {
        // static only before the ‘this(name)’
        ...
        string name = "Test";
        ...
        this(name); // end
    }
    {
        ...
    }
}
```

**Reply by Rekkonnect** [reply]
Why would one go over this hustle and achieve the same thing that constructors inside the method body would enable?

**Comment by cyraid** [comment]
We all just talkin' in the wind again? Does talking here make a difference? How would one actually promote it into an idea the language heads could accept/reject?

**Reply by CyrusNajmabadi** [reply]
Yup. Talking here makes a difference. It's how the process works.

**Reply by cyraid** [reply]
How do we know it makes a difference? Who needs to see it? Where do we vote? Are there other cases where discussions got promoted to implementation?

Thank you Chris and Rekkonnect for the thumbs down for my question.

**Reply by Rekkonnect** [reply]
Thank me too for my downvote.

You are fed up with the lack of direct communication, me too. But nobody prevents me from voicing my proposals, at which point I've done my part and expressed a desire for the evolution of this language, no matter the difficulties of actually influencing and moving the feature into the compiler itself.

**Reply by cyraid** [reply]
Yeah, I hear ya. This used to be an issue, but got locked and thrown into discussions: which is also fine, because then it can be hashed out. But like you said, it's the lack of direct communication that is apparent.

I think a nice voting system, even by number of thumbs (like other projects) so we know something would be considered, but hey. My two cents.

(Also seeing as it's been 4 years since the issue was created)

**Reply by CyrusNajmabadi** [reply]
> Who needs to see it? 

The language design team needs to see and approve it. We do continually watch, engage, and review everything here (I'm one of the design team members). 

> Where do we vote? 

You can vote in the top post, leaving a thumbs up/down/etc.  

> Are there other cases where discussions got promoted to implementation?

Yes. Practically all language changes happen that way.

**Reply by CyrusNajmabadi** [reply]
> This used to be an issue, but got locked and thrown into discussions: 

"Issues" are where we work on approved features, and the design process and implementation tracking.  Discussions are where design and free form thinking happens prior to being approved to be a future language feature. 

> But like you said, it's the lack of direct communication that is apparent.

You are getting direct communication. :-)

> I think a nice voting system, even by number of thumbs (like other projects) so we know something would be considered, 

We consider practically everything submitted to the repo (except things like spam/etc.). But, practically, we only will likely pick a small handful of items per release. 

> Also seeing as it's been 4 years since the issue was created)

Time has no bearing here. We do the work when we think it's the most suitable time. Sometimes that means 10 weeks later, sometimes 10 years :-)

**Reply by HaloFour** [reply]
> This used to be an issue, but got locked and thrown into discussions.

Discussions are a relatively new feature to Github.  Before they were added, all feature requests were created as Issues.  But when Discussions became available it was decided that all requests would be moved to Discussions which have a threaded format, and Issues would br reserved to actual design and spec work for the language team.

**Reply by cyraid** [reply]
Thanks for the reply guys. Answered everything in one swoop, very clear answers too, I like it. I'm used to other projects where when thrown in discussions, it's like a graveyard. Nice to see the mold being broken, keep up the good work.

**Comment by ThaDaVos** [comment]
Bit late - but I think the general idea is to have to same functionality as Java/PHP has with their constructors - that you, the programmer, can decide when the base constructor is called:

**PHP:** https://www.php.net/manual/en/language.oop5.decon.php
**Java:** https://www.javatpoint.com/super-keyword

I also see some use to this - as it allows to transform an input argument for the constructor - or helps in clarity of the code even, as it's easier to read then some multilined `base()` because you have to include some huge argument based initializers for certain parameters (yes this can happen when you use external libraries and extend their class to add functionality and remove duplicate code)

The downside I see of the current workarounds (the out params and static function) - is that you it's freely to move around - so there's nothing preventing it from being moved anywhere except next to the constructor - by having the requested feature of being able to call `base` yourself inside the constructor - the code stays together and stays readable
